---
title: "Standard functions"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
t_six = data.frame(t(sixteenS))
t_meta = data.frame(t(metagenome2))
t_meta$method = "Metagenome"
t_meta$Plant = rownames(t_meta)
t_six$method = "SixteenS"
t_six$Plant = rownames(t_six)
meta_melt = reshape2::melt(t_meta, variable.name = "Family", value.name = "meta_RA",  idvars = c("Plant", "method"))
six_melt = reshape2::melt(t_six, variable.name = "Family", value.name = "six_RA", idvars = c("Plant", "method"))
all_melt = merge(meta_melt, six_melt, by=c("Plant", "Family"))
all_melt$Family[which(all_melt$Family %ni% u)]=as.factor("Other")


comp = ggplot(data = all_melt, aes(x = (six_RA + 0.0001)*100, y = (meta_RA+0.0001)*100, col = Family )) +
  geom_point() + 
  scale_color_manual(values = hue1_25) +
  theme_bw() +
  geom_abline(slope = 1, linetype = "dashed", color = "Grey", size = 1.0) +
  scale_x_log10() +
  scale_y_log10() +
  xlab("16S RA (%)") +
  ylab("Metagenome RA (%)") +
  theme_minimal_hgrid(12, rel_small = 1) +
  theme(
    legend.position = "top",
    legend.justification = "right",
    legend.text = element_text(size = 9),
    legend.box.spacing = unit(0, "pt"),
    legend.title = element_blank()
  ) +
  guides(colour = guide_legend(override.aes = list(size=5)))
  
```